R/teal_lockfile.R

```diff
@@ -23,92 +44,139 @@
 #'
 #' @seealso [renv::snapshot()], [renv::restore()].
 #'
-#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#' @return
+#'  `ExtendedTask` processing `renv` lockfile or `NULL` if lockfile has been provided through options
+#'  (skipping asynchronous process).
+#'
+#' @name teal_lockfile
+#' @rdname teal_lockfile
 #'
 #' @keywords internal
-teal_lockfile <- function() {
+NULL
+
+#' @rdname teal_lockfile
+#' @keywords internal
+teal_lockfile_invoke <- function() {
   lockfile_path <- "teal_app.lock"
+  shiny::onStop(function() file.remove(lockfile_path))
+
   # If user has setup the file, there is no need to compute a new one.
   user_lockfile <- getOption("teal.renv.lockfile", "")
   if (!identical(user_lockfile, "")) {
     if (file.exists(user_lockfile)) {
       file.copy(user_lockfile, lockfile_path)
-      return(invisible(NULL))
+      logger::log_trace('Lockfile set using option "teal.renv.lockfile" - skipping automatic creation.')
+      return(NULL)
     } else {
       stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
     }
   }
 
-  if (!(is_in_test() || is_r_cmd_check())) {
-    old_plan <- future::plan()
-    # If there is already a parallel (non-sequential) backend, reuse it.
-    if (inherits(old_plan, "sequential")) {
-      future::plan(future::multisession, workers = 2)
+  process <- ExtendedTask$new(
+    function(run, lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd) {
+      mirai::mirai(
+        run(lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd),
+        run = run, lockfile_path = lockfile_path, opts = opts, sysenv = sysenv, libpaths = libpaths, wd = wd
+      )
     }
+  )
 
-    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
-    lockfile_task$invoke(close = inherits(old_plan, "sequential"), lockfile_path)
-    logger::log_trace("lockfile creation invoked.")
-  }
+  suppressWarnings({ # 'package:stats' may not be available when loading
+    process$invoke(
+      run = renv_snapshot,
+      lockfile_path = lockfile_path,
+      opts = options(),
```

Comment:
let's take `opts` as a representative but this applies also to `libpaths`, `sysenv` and `wd`

currently this is present in multiple places increasing overall complexity (starting from the bottom):
- argument of `teal:::renv_snapshot()`
- argument of `mirai::mirai` passed to the call
- argument of anonymous function in `ExtendedTask$new`

I think it's a little too much and we would be able to find a way to avoid this. A couple of ideas from my side:
- `mirai::mirai(run = {options(opts); ...; renv_snapshot(lockfile_path)})` - this is to avoid extending `renv_snapshot()`
- `ExtendedTask$new({options(opts); ...; function(lockfile_path){mirai::mirai(renv_snapshot(lockfile_path), lockfile_path))} })`

Reply:
@pawelru it is because `mirai::mirai` way of passing parameters from the current session to the new session that is starts, which is not ideal of course. Or we could pass the whole environment.

<img width="713" alt="image" src="https://github.com/user-attachments/assets/69e22c14-458c-4d64-a14d-b3c1866ad826">

I don't think I understand what's going on in here

<img width="523" alt="image" src="https://github.com/user-attachments/assets/f732e77d-7cb3-4b38-8060-f2f71b06ca7d">

Reply:
https://github.com/insightsengineering/teal/pull/1272/commits/5349788da901388082c561ceddb4a72c744c4746